pkg/cmd/workflow.go

```diff
@@ -0,0 +1,364 @@
+}
+
+func newAddWorkflowCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add [workflow name]",
+		Short: "Add a relay workflow from a local file",
+		Args:  cobra.MaximumNArgs(1),
+		RunE:  doAddWorkflow,
+	}
+
+	cmd.Flags().StringP("file", "f", "", "Path to relay workflow file.")
+
+	return cmd
+}
+
+func doReplaceWorkflow(cmd *cobra.Command, args []string) error {
+	file, err := readFile(cmd)
+
+	if err != nil {
+		return err
+	}
+
+	workflowName, err := getWorkflowName(args)
+
+	if err != nil {
+		return err
+	}
+
+	Dialog.Info("Replacing workflow " + workflowName)
+
+	workflow, werr := Client.GetWorkflow(workflowName)
+
+	if werr != nil {
+		return werr
+	}
+
+	revision, rerr := Client.CreateRevision(workflowName, file)
+
+	if rerr != nil {
+		return rerr
+	}
+
+	wr := model.NewWorkflowRevision(workflow.Workflow, revision.Revision)
+
+	wr.Output(Config)
+
+	Dialog.Infof(`Successfully updated workflow %v
+			
+Updated configuration is visible at %v`,
+		workflowName,
+		format.GuiLink(Config, "/workflow/%v", workflowName),
+	)
+
+	return nil
+}
+
+func newReplaceWorkflowCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "replace [workflow name]",
+		Short: "Replace the configuration of a relay workflow",
+		Args:  cobra.MaximumNArgs(1),
+		RunE:  doReplaceWorkflow,
+	}
+
+	cmd.Flags().StringP("file", "f", "", "Path to relay workflow file.")
+
+	return cmd
+}
+
+func doDeleteWorkflow(cmd *cobra.Command, args []string) error {
+	workflowName, nerr := getWorkflowName(args)
+
+	if nerr != nil {
+		return nerr
+	}
+
+	proceed, cerr := util.Confirm("Are you sure you want to delete this workflow?", Config)
+
+	if cerr != nil {
+		return cerr
+	}
+
+	if !proceed {
+		return nil
+	}
+
+	Dialog.Progress("Deleting workflow...")
+
+	_, err := Client.DeleteWorkflow(workflowName)
+
+	if err != nil {
+		return err
+	}
+
+	Dialog.Info("Workflow successfully deleted")
+
+	return nil
+}
+
+func newDeleteWorkflowCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "delete [workflow name]",
+		Short: "Delete a relay workflow",
+		Args:  cobra.MaximumNArgs(1),
+		RunE:  doDeleteWorkflow,
+	}
+
+	return cmd
+}
+
+func parseParameter(str string) (key, value string) {
+	idx := strings.Index(str, "=")
+
+	// TODO: This behavior will basically silently discard parameters that are
+	// in the wrong format. Should this, like, panic perhaps? Or notify the user that their
+	// parameters are specified incorrectly?
+	if idx < 0 {
+		return
+	}
+
+	key = str[0:idx]
+	value = str[idx+1:]
+	return
+}
+
+func parseParameters(strs []string) map[string]string {
+	res := make(map[string]string)
+
+	for _, str := range strs {
+		key, val := parseParameter(str)
+
+		// value of empty string could, indeed, be a valid parameter.
+		if key != "" {
+			res[key] = val
+		}
+	}
+
+	return res
+}
+
+func doRunWorkflow(cmd *cobra.Command, args []string) error {
+	params, err := cmd.Flags().GetStringArray("parameter")
+
+	if err != nil {
+		panic(err)
```

Comment:
My take is that we should essentially never panic, because of the bad UX it creates in a CLI. I've been using the generic error `return errors.NewGeneralUnknownError().WithCause(err)` instead.

Reply:
We should return errors in scenarios that the issue can be corrected (or will be corrected). In scenarios where you must upgrade the software (e.g. it's a developer error on our end) then panicing is the best bet. We should find all these panics during testing!

Reply:
Errawr has a `.Bug()` you can add to the end to make it (and any errors that follow) buggy. I don't think we should deliberately panic anywhere in this application unless we're going to actually catch the panics and process them in `main()`: users don't care that this software is written in Go, and they certainly won't know what action they should take if they get a stack dump in front of them.

Reply:
My point was more that users should never see a panic. It should be obvious to them that the software has failed catastrophically if they encounter some sort of issue that necessitates a panic. Think like an assertion in C++ land.

On one hand, using the `Bug` feature is handy. One of the conveniences of using the `panic` methodology, though, is that you don't have to deal with certain types of errors. Migrating to the `Bug` feature will make our software more verbose and make it more difficult to differentiate between failure modes that are due to data vs due to the programming.

Reply:
After discussion in meatspace I'm going to revert to using the errawr methodology to conform with our standards everywhere else.